### PR TITLE
Add support for JSON resources to Xliff tasks

### PR DESCRIPTION
**Pull request by tmat**
Implements https://github.com/dotnet/arcade/issues/14275

### CODE REVIEW

**Comment by Cosifne** [comment]
Looks like build is failed?

File: src/Microsoft.DotNet.XliffTasks/Tasks/MetadataKey.cs [hunk 1]
```diff
@@ -13,6 +13,7 @@ internal static class MetadataKey
         public const string XlfLanguage = nameof(XlfLanguage);
         public const string XlfSource = nameof(XlfSource);
         public const string XlfSourceFormat = nameof(XlfSourceFormat);
+        public const string XlfPreserveFileName = nameof(XlfPreserveFileName);
```
Comment (Cosifne) [inline src/Microsoft.DotNet.XliffTasks/Tasks/MetadataKey.cs:16]: What does this refer to?

**Comment by Cosifne** [comment]
(Might be wrong)
What I feel this PR is performing
1. Support load `.json` file from `string-resources.jso`n.
2. it will be expanded to `<lang>.xlf`  in memory.
3. it could later be translated by using the translation dictionary (provided by LOC team).
4. Save as `<lang>.xlf`.

I feel in Step 4 it should be converted to` <lang>/string-resources.json` according the the [doc](https://learn.microsoft.com/en-us/visualstudio/extensibility/visualstudio.extensibility/command/localize-metadata?view=vs-2022) Do I miss anything here?

Might need other's review on details since I am unfamilar with codebase.

File: src/Microsoft.DotNet.XliffTasks/Tasks/MetadataKey.cs (hunk 1 above)
Comment (tmat) [reply src/Microsoft.DotNet.XliffTasks/Tasks/MetadataKey.cs:16]: `XlfPreserveFileName` specifies the naming convention for the translated file.

By default (`XlfPreserveFileName` is `false` or unspecified) the translation process takes an item (e.g. `.resx`) and generates translated `{project name}.xlf\{name}.{lang}.resx` item in the obj directory of the project. For resx the translated file then gets compiled into a satellite assembly.

When `XlfPreserveFileName` is set to true on a translatable item it specifies that instead of inserting the language name (`{lang}`) into the file name we preserve the file name and store it in a `{lang}` subdirectory. E.g. `{lang}\string-resources.json`.

Comment (tmat) [reply src/Microsoft.DotNet.XliffTasks/Tasks/MetadataKey.cs:16]: I'll add a comment

File: src/Microsoft.DotNet.XliffTasks/build/Microsoft.DotNet.XliffTasks.targets [hunk 2]
```diff
@@ -2,8 +2,7 @@
 <Project>
 
   <PropertyGroup>
-    <XliffTasksDirectory Condition="'$(XliffTasksDirectory)' == '' and '$(MSBuildRuntimeType)' == 'Core'">$(MSBuildThisFileDirectory)..\tools\net9.0\</XliffTasksDirectory>
-    <XliffTasksDirectory Condition="'$(XliffTasksDirectory)' == '' and '$(MSBuildRuntimeType)' != 'Core'">$(MSBuildThisFileDirectory)..\tools\net472\</XliffTasksDirectory>
+    <XliffTasksDirectory Condition="'$(XliffTasksDirectory)' == ''">$(MSBuildThisFileDirectory)..\tools\$(MSBuildRuntimeType)\</XliffTasksDirectory>
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.XliffTasks/build/Microsoft.DotNet.XliffTasks.targets:5]: I think it would be better (safer) to not rely on $(MSBuildRuntimeType) here. Also, I would use other terms than `Full` and `Core` as those don't reflect the current framework names anymore.

```suggestion
    <XliffTasksDirectory Condition="'$(XliffTasksDirectory)' == '' and '$(MSBuildRuntimeType)' == 'Core'">$(MSBuildThisFileDirectory)..\tools\net\</XliffTasksDirectory>
    <XliffTasksDirectory Condition="'$(XliffTasksDirectory)' == '' and '$(MSBuildRuntimeType)' != 'Core'">$(MSBuildThisFileDirectory)..\tools\netframework\</XliffTasksDirectory>
```

File: src/Microsoft.DotNet.XliffTasks/Microsoft.DotNet.XliffTasks.csproj [hunk 3]
```diff
@@ -1,11 +1,19 @@
...
+    -->
+    <TargetFrameworks>$(NetToolMinimum);$(NetFrameworkToolCurrent)</TargetFrameworks>
     <IsPackable>true</IsPackable>
     <RootNamespace>XliffTasks</RootNamespace>
     <StrongNameKeyId>MicrosoftAspNetCore</StrongNameKeyId>
     <PackageReadmeFile>README.md</PackageReadmeFile>
+
+    <!-- The subdirectory name is the value of $(MSBuildRuntimeType). -->
+    <BuildTaskTargetTfmSpecificFolder>tools/Core</BuildTaskTargetTfmSpecificFolder>
+    <BuildTaskTargetTfmSpecificFolder Condition="'$(TargetFramework)' == '$(NetFrameworkToolCurrent)'">tools/Full</BuildTaskTargetTfmSpecificFolder>
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.XliffTasks/Microsoft.DotNet.XliffTasks.csproj:16]: Use a condition in L15 to avoid the property redefine when compiling against .NET Framework. Also use better framework names:

```suggestion
    <BuildTaskTargetTfmSpecificFolder Condition="'$(TargetFramework)' == '$(NetToolMinimum)'">tools/net</BuildTaskTargetTfmSpecificFolder>
    <BuildTaskTargetTfmSpecificFolder Condition="'$(TargetFramework)' == '$(NetFrameworkToolCurrent)'">tools/netframework</BuildTaskTargetTfmSpecificFolder>
```

File: eng/BuildTask.targets [hunk 4]
```diff
@@ -84,18 +84,24 @@
                Pack="false" />
   </ItemGroup>
 
-  <!-- Publish .NET Core assets and include them in the package under $(BuildTaskTargetFolder) directory. -->
+  <!-- Publish .NET Core assets and include them in the package under $(BuildTaskTargetTfmSpecificFolder) directory. -->
   <Target Name="_AddBuildOutputToPackageCore" DependsOnTargets="Publish" Condition="'$(TargetFrameworkIdentifier)' == '.NETCoreApp'">
+    <PropertyGroup>
+      <BuildTaskTargetTfmSpecificFolder Condition="'$(BuildTaskTargetTfmSpecificFolder)' == ''">$(BuildTaskTargetFolder)/($TargetFramework)</BuildTaskTargetTfmSpecificFolder>
```
Comment (ViktorHofer) [inline eng/BuildTask.targets:90]: We already have a property for the root level folder `BuildTaskTargetFolder`. Continue to respect that:

```suggestion
      <BuildTaskTargetTfmSpecificFolder Condition="'$(BuildTaskTargetTfmSpecificFolder)' == ''">($TargetFramework)</BuildTaskTargetTfmSpecificFolder>
```

File: eng/BuildTask.targets [hunk 5]
```diff
@@ -84,18 +84,24 @@
...
+    </PropertyGroup>
     <ItemGroup>
       <TfmSpecificPackageFile Include="$(PublishDir)**"
-                              PackagePath="$(BuildTaskTargetFolder)/$(TargetFramework)/%(RecursiveDir)%(FileName)%(Extension)"/>
+                              PackagePath="$(BuildTaskTargetTfmSpecificFolder)/%(RecursiveDir)%(FileName)%(Extension)"/>
```
Comment (ViktorHofer) [inline eng/BuildTask.targets:94]:
```suggestion
                              PackagePath="$(BuildTaskTargetFolder)/$(BuildTaskTargetTfmSpecificFolder)/%(RecursiveDir)%(FileName)%(Extension)"/>
```

File: eng/BuildTask.targets [hunk 6]
```diff
@@ -84,18 +84,24 @@
...
     </ItemGroup>
   </Target>
 
-  <!-- Include .NET Framework build outputs in the package under $(BuildTaskTargetFolder) directory. -->
+  <!-- Include .NET Framework build outputs in the package under $(BuildTaskTargetTfmSpecificFolder) directory. -->
   <Target Name="_AddBuildOutputToPackageDesktop" Condition="'$(TargetFrameworkIdentifier)' != '.NETCoreApp'">
+    <PropertyGroup>
+      <BuildTaskTargetTfmSpecificFolder Condition="'$(BuildTaskTargetTfmSpecificFolder)' == ''">$(BuildTaskTargetFolder)/($TargetFramework)</BuildTaskTargetTfmSpecificFolder>
```
Comment (ViktorHofer) [inline eng/BuildTask.targets:101]:
```suggestion
      <BuildTaskTargetTfmSpecificFolder Condition="'$(BuildTaskTargetTfmSpecificFolder)' == ''">$(TargetFramework)</BuildTaskTargetTfmSpecificFolder>
```

File: eng/BuildTask.targets [hunk 7]
```diff
@@ -84,18 +84,24 @@
...
+    </PropertyGroup>
     <ItemGroup>
-      <TfmSpecificPackageFile Include="$(OutputPath)**" PackagePath="$(BuildTaskTargetFolder)/$(TargetFramework)/%(RecursiveDir)%(FileName)%(Extension)"/>
+      <TfmSpecificPackageFile Include="$(OutputPath)**" PackagePath="$(BuildTaskTargetTfmSpecificFolder)/%(RecursiveDir)%(FileName)%(Extension)"/>
```
Comment (ViktorHofer) [inline eng/BuildTask.targets:104]:
```suggestion
      <TfmSpecificPackageFile Include="$(OutputPath)**" PackagePath="$(BuildTaskTargetFolder)/$(BuildTaskTargetTfmSpecificFolder)/%(RecursiveDir)%(FileName)%(Extension)"/>
```

File: Directory.Packages.props [hunk 8]
```diff
@@ -50,21 +50,34 @@
...
+    <PackageVersion Include="System.Reflection.Metadata" Version="$(SystemReflectionMetadataVersion)" />
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(TargetFramework)' == 'net472'" >
+    <PackageVersion Include="System.Text.Json" Version="7.0.1" />
+    <PackageVersion Include="Microsoft.Extensions.Logging.Console" Version="7.0.0" />
+    <PackageVersion Include="Microsoft.Extensions.DependencyModel" Version="7.0.0" />
+    <PackageVersion Include="System.Collections.Immutable" Version="7.0.0" />
+    <PackageVersion Include="System.Reflection.Metadata" Version="7.0.0" />
+  </ItemGroup>
```
Comment (ViktorHofer) [inline Directory.Packages.props:79]: This looks wrong to me. .NET Framework libraries that aren't msbuild tasks don't have that restriction.

Comment (ViktorHofer) [reply Directory.Packages.props:79]: I think it's better to revert the changes in this file (I also don't like the TFM condition here) and use `VersionOverride` in the msbuild package libraries.

Comment (ViktorHofer) [reply Directory.Packages.props:79]: In addition, these versions should be `8.0.x` as roslyn, msbuild and others just updated to them.

Comment (tmat) [reply Directory.Packages.props:79]: 17.9.0 Preview 2.0 still has 7.0.0 and CI machines have Preview 1.0, so can't update to 8.0.0 yet.

Comment (tmat) [reply Directory.Packages.props:79]: I'll use VersionOverride, but in general we only target net472 for build tasks. I don't see other reasons why we would target it.

Comment (ViktorHofer) [reply Directory.Packages.props:79]: We have libraries like XUnitExtensions, RemoteExecutor and others in this repository that aren't msbuild task. Those currently target a lower TFM but this might change in the future. Globally conditioning these versions on a specific TFM is wrong. This is about the type of project, not the TFM.

Comment (ViktorHofer) [reply Directory.Packages.props:79]: VersionOverride on the PackageReferences should work. It might be easiest to either add the `VersionOverride` metadata to the PackageReferences that are already are in https://github.com/dotnet/arcade/blob/main/eng/BuildTask.targets or you could add `PackageVersion` update statements in that file.

Comment (tmat) [reply Directory.Packages.props]: OK, this works: We set property `IsBuildTaskProject` in each build task project and condition the package version based on it. I have updated Xliff project. If it looks good I'll update all other build task projects.

File: src/Microsoft.DotNet.Arcade.Sdk/tools/TargetFrameworkDefaults.props [hunk 9]
```diff
@@ -37,8 +37,9 @@
     <NetToolCurrent Condition="'$(DotNetBuildFromSource)' == 'true'">$(NetCurrent)</NetToolCurrent>
 
     <!-- Lowest version of .NET at the time of the release of NetCurrent that is supported by tooling.
-         Undefined when NetToolCurrent and NetToolMinimum are identical. -->
-    <NetToolMinimum>net8.0</NetToolMinimum>
+         Identical with NetCurrent when building from source. -->
+    <NetToolMinimum Condition="'$(DotNetBuildFromSource)' != 'true'">$(NetMinimum)</NetToolMinimum>
+    <NetToolMinimum Condition="'$(DotNetBuildFromSource)' == 'true'">$(NetCurrent)</NetToolMinimum>
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.Arcade.Sdk/tools/TargetFrameworkDefaults.props:42]: We don't want NetToolMinimum to default to NetMinimum. We version those two properties differently in some cases. Basically NetTool* properties shouldn't rely on Net* properties.

```suggestion
         Identical with NetToolCurrent when building from source. -->
    <NetToolMinimum Condition="'$(DotNetBuildFromSource)' != 'true'">net8.0</NetToolMinimum>
    <NetToolMinimum Condition="'$(DotNetBuildFromSource)' == 'true'">$(NetToolCurrent)</NetToolMinimum>
```

File: Directory.Build.targets [hunk 10]
```diff
@@ -2,5 +2,6 @@
 
   <Import Project="Sdk.targets" Sdk="Microsoft.DotNet.Arcade.Sdk" />
   <Import Project="$(RepositoryEngineeringDir)Microsoft.DotNet.XliffTasks.InTree.targets" Condition="'$(UsingInTreeToolXliff)' == 'true'" />
+  <Import Project="$(RepositoryEngineeringDir)BuildTask.targets" Condition="'$(IsBuildTaskProject)' == 'true'" />
```
Comment (ViktorHofer) [inline Directory.Build.targets:5]: I like this concept. All the other task projects in this repo need to be updated as well though. If you don't plan to do this as part of this PR, please file an issue to track that work.

File: src/Microsoft.DotNet.XliffTasks/Microsoft.DotNet.XliffTasks.csproj [hunk 11]
```diff
@@ -1,11 +1,20 @@
...
+       Xliff tasks were moved from dotnet/xliff-tasks to Arcade in V9, so they are not available in release/8.0 branch.
+    -->
+    <TargetFrameworks>$(NetToolMinimum);$(NetFrameworkToolCurrent)</TargetFrameworks>
     <IsPackable>true</IsPackable>
+    <IsBuildTaskProject>true</IsBuildTaskProject>
     <RootNamespace>XliffTasks</RootNamespace>
     <StrongNameKeyId>MicrosoftAspNetCore</StrongNameKeyId>
     <PackageReadmeFile>README.md</PackageReadmeFile>
+
+    <!-- The subdirectory name is the value of $(MSBuildRuntimeType). -->
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.XliffTasks/Microsoft.DotNet.XliffTasks.csproj:15]:
```suggestion
```

File: src/Microsoft.DotNet.XliffTasks/Microsoft.DotNet.XliffTasks.csproj [hunk 12]
```diff
@@ -1,11 +1,20 @@
...
+    <BuildTaskTargetTfmSpecificFolder>tools/net</BuildTaskTargetTfmSpecificFolder>
+    <BuildTaskTargetTfmSpecificFolder Condition="'$(TargetFramework)' == '$(NetFrameworkToolCurrent)'">tools/netframework</BuildTaskTargetTfmSpecificFolder>
```
Comment (ViktorHofer) [inline src/Microsoft.DotNet.XliffTasks/Microsoft.DotNet.XliffTasks.csproj:17]: This should eventually be consolidated with all the other task projects in this repository. Please file an issue for that.

File: Directory.Packages.props [hunk 13]
```diff
@@ -50,21 +50,34 @@
...
+    <PackageVersion Include="System.Reflection.Metadata" Version="$(SystemReflectionMetadataVersion)" />
+  </ItemGroup>
+
+  <ItemGroup Condition="'$(TargetFramework)' == 'net472' and '$(IsBuildTaskProject)' == 'true'" >
+    <PackageVersion Include="System.Text.Json" Version="7.0.1" />
+    <PackageVersion Include="Microsoft.Extensions.Logging.Console" Version="7.0.0" />
+    <PackageVersion Include="Microsoft.Extensions.DependencyModel" Version="7.0.0" />
+    <PackageVersion Include="System.Collections.Immutable" Version="7.0.0" />
+    <PackageVersion Include="System.Reflection.Metadata" Version="7.0.0" />
+  </ItemGroup>
```
Comment (ViktorHofer) [inline Directory.Packages.props:79]: This one is my only remaining feedback that I feel very strongly about. I think all the changes in this file should be reverted. The overrides (i.e. via `<PackageVersion Update="..." />`) should happen in BuildTasks.targets.

File: src/Microsoft.DotNet.XliffTasks/Microsoft.DotNet.XliffTasks.csproj (hunk 12 above)
Comment (tmat) [reply src/Microsoft.DotNet.XliffTasks/Microsoft.DotNet.XliffTasks.csproj:16]: Yes, we should remove this property and just do it for all build tasks.

File: Directory.Packages.props (hunk 13 above)
Comment (tmat) [reply Directory.Packages.props:79]: I can try that.

File: Directory.Build.targets (hunk 10 above)
Comment (tmat) [reply Directory.Build.targets:5]: Done.

File: src/Microsoft.DotNet.XliffTasks/Microsoft.DotNet.XliffTasks.csproj (hunk 12 above)
Comment (tmat) [reply src/Microsoft.DotNet.XliffTasks/Microsoft.DotNet.XliffTasks.csproj:16]: https://github.com/dotnet/arcade/issues/14417

File: Directory.Packages.props (hunk 13 above)
Comment (tmat) [reply Directory.Packages.props]: Done.

File: Directory.Packages.props [hunk 14]
```diff
@@ -50,17 +50,18 @@
...
     <PackageVersion Include="System.IO.Packaging" Version="$(SystemIOPackagingVersion)" />
-    <PackageVersion Include="System.Reflection.Metadata" Version="$(SystemReflectionMetadataVersion)" />
     <PackageVersion Include="System.Security.Cryptography.Xml" Version="$(SystemSecurityCryptographyXmlVersion)" />
     <PackageVersion Include="System.Text.Encodings.Web" Version="$(SystemTextEncodingsWebVersion)" />
     <PackageVersion Include="System.Text.Json" Version="$(SystemTextJsonVersion)" />
+    <PackageVersion Include="Microsoft.Extensions.Logging.Console" Version="$(MicrosoftExtensionsLoggingConsoleVersion)" />
+    <PackageVersion Include="Microsoft.Extensions.DependencyModel" Version="$(MicrosoftExtensionsDependencyModelVersion)" />
+    <PackageVersion Include="System.Collections.Immutable" Version="$(SystemCollectionsImmutableVersion)" />
+    <PackageVersion Include="System.Reflection.Metadata" Version="$(SystemReflectionMetadataVersion)" />
+
```
Comment (ViktorHofer) [inline Directory.Packages.props:64]: nit: undo these changes

File: eng/BuildTask.Packages.props [hunk 15]
```diff
@@ -0,0 +1,9 @@
+<Project>
+  <ItemGroup Condition="'$(TargetFramework)' == 'net472'">
```
Comment (ViktorHofer) [inline eng/BuildTask.Packages.props:2]: This be conditioned on the TFM family not the exact TFM:

```suggestion
  <ItemGroup Condition="'$(TargetFrameworkIdentifier)' == '.NETFramework'">
```

File: src/Common/Microsoft.Arcade.Common/Microsoft.Arcade.Common.csproj [hunk 16]
```diff
@@ -8,7 +10,7 @@
   </PropertyGroup>
 
   <ItemGroup>
-    <PackageReference Include="Microsoft.Build.Utilities.Core" />
+    <PackageReference Include="Microsoft.Build.Utilities.Core" VersionOverride="" />
```
Comment (ViktorHofer) [inline src/Common/Microsoft.Arcade.Common/Microsoft.Arcade.Common.csproj:13]: undo?

```suggestion
    <PackageReference Include="Microsoft.Build.Utilities.Core" />
```

File: eng/BuildTask.targets [hunk 17]
```diff
@@ -1,4 +1,5 @@
 <Project>
+  <Import Project="BuildTask.Packages.props"/>
```
Comment (ViktorHofer) [inline eng/BuildTask.targets:2]:
```suggestion

  <Import Project="BuildTask.Packages.props"/>
```

File: eng/BuildTask.Packages.props (hunk 15 above)
Comment (ViktorHofer) [inline eng/BuildTask.Packages.props:2]:
```suggestion

  <ItemGroup Condition="'$(TargetFramework)' == 'net472'">
```

File: eng/BuildTask.Packages.props [hunk 18]
```diff
@@ -0,0 +1,9 @@
...
+    <PackageVersion Update="System.Text.Json" Version="7.0.1" />
+    <PackageVersion Update="Microsoft.Extensions.Logging.Console" Version="7.0.0" />
+    <PackageVersion Update="Microsoft.Extensions.DependencyModel" Version="7.0.0" />
+    <PackageVersion Update="System.Collections.Immutable" Version="7.0.0" />
+    <PackageVersion Update="System.Reflection.Metadata" Version="7.0.0" />
+  </ItemGroup>
+</Project>
```
Comment (ViktorHofer) [inline eng/BuildTask.Packages.props:9]:
```suggestion

</Project>
```

**Review by ViktorHofer — APPROVED** [review]

**Review by ViktorHofer — APPROVED** [review]

**Review by Cosifne — APPROVED** [review]
